src/mscorlib/shared/System/Threading/ReaderWriterLockSlim.cs

```diff
@@ -1059,33 +1121,273 @@ private uint GetNumReaders()
             return _owners & READER_MASK;
         }
 
+        private bool ShouldSpinForEnterAnyRead()
+        {
+            // If there is a write waiter or write upgrade waiter, the waiter would block a reader from acquiring the RW lock
+            // because the waiter takes precedence. In that case, the reader is not likely to make progress by spinning.
+            return _fNoWaiters || _numWriteWaiters == 0 && _numWriteUpgradeWaiters == 0;
+        }
+
+        private bool ShouldSpinForEnterAnyWrite(bool isUpgradeToWrite)
+        {
+            // If there is a write upgrade waiter, the waiter would block a writer from acquiring the RW lock because the waiter
+            // holds a read lock. In that case, the writer is not likely to make progress by spinning.
+            return isUpgradeToWrite || _numWriteUpgradeWaiters == 0;
+        }
+
+        private uint EnterMyLockForEnterAnyReadDeprioritizedCount => (uint)((ulong)_enterMyLockDeprioritizationState >> 32);
+        private uint EnterMyLockForEnterAnyWriteDeprioritizedCount => (uint)_enterMyLockDeprioritizationState;
+
+        private static void WaitOnceForEnterMyLock(int spinIndex, int processorCount)
+        {
+            if (spinIndex < LockSpinCount && processorCount > 1)
+            {
+                RuntimeThread.SpinWait(LockSpinCycles * (spinIndex + 1)); // Wait a few dozen instructions to let another processor release lock.
+            }
+            else if (spinIndex < (LockSpinCount + LockSleep0Count))
+            {
+                RuntimeThread.Sleep(0);   // Give up my quantum.
+            }
+            else
+            {
+                RuntimeThread.Sleep(1);   // Give up my quantum.
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool TryEnterMyLock()
+        {
+            return Interlocked.CompareExchange(ref _myLock, 1, 0) == 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void EnterMyLockForEnterAnyRead(bool isForWait = false)
+        {
+            if (!TryEnterMyLock())
```

Comment:
Good point, when the enter is deprioritized it would mean there is no progress to be made on the RW lock so it doesn't have to spin for _myLock. I'll also fix it so that it doesn't create/wait on the event when it decides to skip spinning on the RW lock and the timeout has expired.